### PR TITLE
Separate out unit test projects

### PR DESCRIPTION
**Pull request by jaredpar**
Unit test projects are a kind of deployment project.  This means they need to deploy as a part of build in order to satisfy our core scenarios around F5 and right click test.  

Deployment projects need to have a unique output directory in order to have correct build semantics.  Building into a shared directory means dependencies are overwritten many times during build and open the door to build race conditions.  Today though our unit tests build into a shared directory due to the circumstances of our build setup:

1. Several unit test projects reference each other. 
2. All unit test runners look for *.UnitTests.dll under a certain directory. 

If our unit tests built into unique directories then all of our test runners would end up finding two copies of some tests and hence end up double testing.  For example Microsoft.CodeAnalysis.Scripting.UnitTests.dll would be deployed twice under our UnitTest directory and hence found twice by the test runner search code. 

One of the above issues needs to be addressed in order to have a correct build environment for Roslyn.  I gave a very long look at addressing the second item.  However I could not come up with a way to implement this simply and correctly.  Additionally looking around at other repos virtually every one solves at as we do today with a simple search.  Felt like I was swimming upstream on this.

The first issue is addressable by moving the shared code to a helper DLL and removing the now unnecessary references.  In most cases there was a non-test project readily available to hold the code.  In other cases, like Scripting, a new one needed to be introduced.  This is a bit of labor but very approachable and hence I chose this approach. 

Note: This fixes our unit test up through compiler and scripting .  Future changes will address workspaces, features and Visual Studio.

### CODE REVIEW

**Comment by jaredpar** [comment]
CC @tmat  for the scripting test change (it's the largest) 
CC @dotnet/roslyn-compiler  for compiler changes.  This is just moving code to a shared DLL and deleting a few references though.

**Comment by jaredpar** [comment]
CC @dotnet/roslyn-ide as this change now encompasses fixing the IDE unit tests as well.

There is a lot more apparent churn to fix the IDE tests.  Really though 99% of it is just moving a bunch of types from EditorServicesTests to ServiceTestUtilities.  The code didn't change, just the location.

File: src/Workspaces/CoreTestUtilities/Formatting/FormattingTestBase.cs [hunk 1]
```diff
@@ -21,7 +21,7 @@ public abstract class FormattingTestBase
             Dictionary<OptionKey, object> changedOptionSet = null,
             bool testWithTransformation = true)
         {
-            return AssertFormatAsync(expected, code, SpecializedCollections.SingletonEnumerable(new TextSpan(0, code.Length)), language, debugMode, changedOptionSet, testWithTransformation);
+            return AssertFormatAsync(expected, code, new[] { new TextSpan(0, code.Length)}, language, debugMode, changedOptionSet, testWithTransformation);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CoreTestUtilities/Formatting/FormattingTestBase.cs]: spacing.

Comment (CyrusNajmabadi) [inline src/Workspaces/CoreTestUtilities/Formatting/FormattingTestBase.cs]: Also, why isn't this using SPecializedCollectoins anymore?

**Comment by tmat** [comment]
👍

**Comment by CyrusNajmabadi** [comment]
That would be awesome.  can't wait for this to go in!

**Comment by jaredpar** [comment]
@Pilchie, @CyrusNajmabadi there is still one more change needed to 100% separate out the projects.  It involves a small amount of actual code change though so I wanted to review it separately.  Dont' think it's a lot though and hope to be done EOD.

File: src/Workspaces/CoreTestUtilities/Formatting/FormattingTestBase.cs (hunk 1 above)
Comment (jaredpar) [inline src/Workspaces/CoreTestUtilities/Formatting/FormattingTestBase.cs]: The `SpecializedCollections` type came through an IVT attribute.  Once the code was moved it was difficult to keep without causing other IVT conflicts.  Hence I just simplified and used an array.

If this is an important perf point I can work to fix this in a later check-in.

**Comment by jaredpar** [comment]
Note: the ETA test is going to fail with a BuildBoss validation error.  That is expected.  Once I get this merged I have a closed change that will fix that up.

**Comment by agocke** [comment]
LGTM

File: src/EditorFeatures/Test/SolutionCrawler/WorkCoordinatorTests.cs [hunk 2]
```diff
@@ -825,7 +825,7 @@ End Property
         private async Task InsertText(string code, string text, bool expectDocumentAnalysis, string language = LanguageNames.CSharp)
         {
```
Comment (jaredpar) [inline src/EditorFeatures/Test/SolutionCrawler/WorkCoordinatorTests.cs]: @CyrusNajmabadi @Pilchie wanted to make sure you saw the changes I made in this file. 

This test defined two exportable MEF components.  In the past they were loaded into the majority of composition due to a quirk of our loading code. A lot of our code in `TestExportProvider` specifies types which should load yet there are a lot of consumers which just convert all of those `Type` values into `Type.Assembly`.  Hence really when we say "type" should load we're saying "nah just load everything in this assembly".  

In the past this caused `WorkspaceWaiter` and `SolutionCrawlerWaiter` to be loaded into most compositions.  Once the assemblies split this stopped being true.  Had to create a custom `ExportProvider` for this test to get it working again.

**Comment by jaredpar** [comment]
@CyrusNajmabadi @Pilchie with the last set of commits the unit tests are now fully broken apart.  Every unit test project is now a leaf node in our build graph.  So should be fair game on whatever was blocking you for Dev15 once I get this merged + the change to actually build to separate dirs.  That will be a sep change though because I'm betting we discover some more implicit dependency bugs when that happens.

**Comment by jaredpar** [comment]
@CyrusNajmabadi do u want me to hold up the change until we can discuss the waiter issue? Happy to merge and do more clean up tomorrow if there turns out to be a better way.

**Comment by CyrusNajmabadi** [comment]
no.  def do tnot hold the change.  :)

**Comment by jaredpar** [comment]
Thanks.  Merge all the things!!!

**Pull request merged**